inferno-ml-server-types/src/Inferno/ML/Server/Types.hs

```diff
@@ -0,0 +1,762 @@
+    outputs :: Vector (SingleOrMany p),
+    user :: uid
+  }
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData)
+
+-- We only want this instance if the `script` is a `VCObjectHash` (because it
+-- should not be possible to store a new param with a raw script)
+instance
+  ( FromJSON p,
+    Typeable p,
+    FromField uid
+  ) =>
+  FromRow (InferenceParam uid gid p VCObjectHash)
+  where
+  fromRow =
+    InferenceParam
+      <$> field
+      <*> fmap wrappedTo (field @VCObjectHashRow)
+      <*> field
+      -- HACK / FIXME This is a pretty awful hack (storing as `jsonb`),
+      -- but Postgres sub-arrays need to be the same length and writing
+      -- a custom parser might be painful
+      <*> fmap getAeson field
+      <*> fmap getAeson field
+      <*> field
+
+instance
+  ( ToJSON p,
+    ToField uid
+  ) =>
+  ToRow (InferenceParam uid gid p VCObjectHash)
+  where
+  -- NOTE: Do not change the order of the field actions
+  toRow ip =
+    [ toField Default,
+      ip ^. the @"script" & VCObjectHashRow & toField,
+      ip ^. the @"model" & toField,
+      -- HACK / FIXME See above
+      ip ^. the @"inputs" & Aeson & toField,
+      ip ^. the @"outputs" & Aeson & toField,
+      ip ^. the @"user" & toField
+    ]
+
+-- | A user, parameterized by the user and group types
+data User uid gid = User
+  { id :: uid,
+    groups :: Vector gid
+  }
+  deriving stock (Show, Generic, Eq)
+  deriving anyclass (FromRow, ToRow, NFData)
+
+-- | IPv4 address with some useful instances
+newtype IPv4 = IPv4 Data.IP.IPv4
+  deriving stock (Generic)
+  deriving newtype (Show, Eq, Ord, Read)
+
+instance NFData IPv4 where
+  rnf = rwhnf
+
+instance FromJSON IPv4 where
+  parseJSON =
+    withText "IPv4" $
+      maybe (fail "Invalid IPv4") (pure . IPv4)
+        . readMaybe
+        . Text.unpack
+
+instance ToJSON IPv4 where
+  toJSON = String . Text.pack . show
+
+instance FromHttpApiData IPv4 where
+  parseUrlPiece = maybe (Left "Invalid IPv4") Right . readMaybe . Text.unpack
+
+instance ToHttpApiData IPv4 where
+  toUrlPiece = Text.pack . show
+
+instance FromField IPv4 where
+  fromField = maybeConversion $ readMaybe . ByteString.Char8.unpack
+
+instance ToField IPv4 where
+  toField = Escape . ByteString.Char8.pack . show
+
+toIPv4 :: (Int, Int, Int, Int) -> IPv4
+toIPv4 (a, b, c, d) = IPv4 $ Data.IP.toIPv4 [a, b, c, d]
+
+fromIPv4 :: IPv4 -> (Int, Int, Int, Int)
+fromIPv4 =
+  wrappedTo >>> Data.IP.fromIPv4 >>> \case
+    [a, b, c, d] -> (a, b, c, d)
+    -- Should not happen
+    _ -> error "Invalid IP address"
+
+-- Bridge-related types
+
+-- | A value that can be used with Inferno and that will be consumed or returned
+-- by @inferno-ml-server@ or the bridge server
+--
+-- Note that this is significantly more restrictive than Inferno's @Value@ type,
+-- which cannot have sensible @ToJSON@ and @FromJSON@ instances
+data IValue
+  = IText Text
+  | IDouble Double
+  | ITuple (IValue, IValue)
+  | ITime EpochTime
+  | IEmpty
```

Comment:
We might also want booleans. Also, perhaps we should sync this with the inferno-tachdb value conversions in plow-inferno, to make sure both support the same set of value types

Reply:
Yes, I agree. We should probably add a type representing a limited subset of Inferno values to one of the OSS packages and use that in both places